src/analysis/retail/shaman/elemental/modules/talents/Stormkeeper.tsx

```diff
@@ -1,37 +1,264 @@
+  lastSKHardcast: CastEvent | null = null;
+
+  stSpender: Talent;
+
   damageDoneByBuffedCasts = 0;
 
   constructor(options: Options) {
-    super(options);
+    super({ spell: TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT }, options);
+
+    this.stSpender = this.selectedCombatant.hasTalent(TALENTS.ELEMENTAL_BLAST_TALENT)
+      ? TALENTS.ELEMENTAL_BLAST_TALENT
+      : TALENTS.EARTH_SHOCK_TALENT;
     this.active =
       this.selectedCombatant.getRepeatedTalentCount(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT) > 0;
     if (!this.active) {
       return;
     }
+
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER), this.onEventDuringSK);
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT),
+      this.onSKCast,
+    );
     this.addEventListener(
-      Events.damage.by(SELECTED_PLAYER).spell(AFFECTED_ABILITIES),
+      Events.removebuff.by(SELECTED_PLAYER).spell(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT),
+      this.onSKFalloff,
+    );
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER).spell(SK_DAMAGE_AFFECTED_ABILITIES),
       this.onSKDamage,
     );
   }
 
+  onSKCast(cast: CastEvent) {
+    this.lastSKHardcast = cast;
+  }
+
+  onSKFalloff(event: RemoveBuffEvent) {
+    if (!this.activeWindow) {
+      return;
+    }
+
+    this.activeWindow.timeline.end = event.timestamp;
+
+    this.recordCooldown(this.activeWindow);
+    this.lastSKHardcast = null;
+    this.activeWindow = null;
+  }
+
+  onEventDuringSK(event: CastEvent) {
+    /* Don't include events that did not happen between hardcasts and the
+    SK buff falling off. */
+    if (!this.lastSKHardcast) {
+      return;
+    }
+
+    if (!this.activeWindow && WINDOW_START_SPELLS.includes(event.ability.guid)) {
+      this.activeWindow = {
+        event: this.lastSKHardcast,
+        maelstromOnCast: this.maelstromTracker.current,
+        flameshockDurationOnCast: this.enemies.getLongestDurationRemaining(
+          SPELLS.FLAME_SHOCK.id,
+          FLAMESHOCK_BASE_DURATION,
+          FLAMESHOCK_BASE_DURATION * (1 + PANDEMIC_WINDOW),
+          event.timestamp,
+        ),
+        electrifiedShocksDurationOnCast: this.enemies.getLongestDurationRemaining(
+          SPELLS.ELECTRIFIED_SHOCKS_DEBUFF.id,
+          ELECTRIFIED_SHOCKS_DURATION,
+          ELECTRIFIED_SHOCKS_DURATION,
+          event.timestamp,
+        ),
+        sopOnCast: this.selectedCombatant.hasBuff(SPELLS.SURGE_OF_POWER_BUFF.id),
+        moteOnCast: this.selectedCombatant.hasBuff(SPELLS.MASTER_OF_THE_ELEMENTS_BUFF.id),
+        timeline: {
+          start: this.lastSKHardcast.timestamp,
+          end: -1,
+          events: [this.lastSKHardcast],
+          performance: QualitativePerformance.Perfect,
+        },
+        firstRotationCast: event,
+      };
+
+      /* If the user started the window with a spender, then the maelstrom
+      has already been consumed */
+      if (event.ability.guid === this.stSpender.id) {
```

Comment:
[Do] I think this would read a lot better as a conditional in the object above.  Being down here makes it confusing because figuring out what the value of that field will be requires looking in two different places.  Additionally, can't we just look up the resource cost for any of them?  It'll just be missing if it didn't use maelstrom.

```
maelstromOnCast: this.maelstromTracker.current + (event.resourceCost[RESOURCE_TYPES.MAELSTROM.id] ?? 0),
```

If that doesn't work you can use a ternary expression like
```
maelstromOnCast: this.maelstromTracker.current +
  (event.ability.guid === this.stSpender.id && event.resourceCost && event.resourceCost[RESOURCE_TYPES.MAELSTROM.id])
    ? event.resourceCost[RESOURCE_TYPES.MAELSTROM.id]
    : 0,
```

Reply:
Went with the first suggestion.